Fix #30690: Preserve newlines in import_yaml multi-line scalars

dwoz opened the pull request:
When using import_yaml to load YAML data with multi-line scalars (using |), the newlines were being rendered as literal \n instead of actual newlines when passed through Jinja templates.

This fix adds an _unescape_newlines function that recursively converts literal \n sequences back to actual newlines in the context dictionary before it's used for templating.

- Added _unescape_newlines function to salt/utils/templates.py
- Applied unescape to context before creating decoded_context
- Updated test to remove to_dict filter reference

Fixes #30690

dwoz committed:
Fix #30690: Preserve newlines in import_yaml multi-line scalars

When using import_yaml to load YAML data with multi-line scalars (using |),
the newlines were being rendered as literal \n instead of actual newlines
when passed through Jinja templates.

This fix adds an _unescape_newlines function that recursively converts
literal \n sequences back to actual newlines in the context dictionary
before it's used for templating.

- Added _unescape_newlines function to salt/utils/templates.py
- Applied unescape to context before creating decoded_context
- Updated test to remove to_dict filter reference

twangboy commented on salt/utils/templates.py:
```diff
@@ -428,6 +428,25 @@ def opt_jinja_env_helper(opts, optname):
+                return [_unescape_newlines(item) for item in data]
+            elif isinstance(data, str):
+                # Convert literal \n (backslash-n) to actual newline
+                return data.replace("\\n", "\n")
```
What happens on Windows? Does it need to also look for `\\r\\n`? Maybe need to use `os.linesep` instead?

dwoz replied:
Good catch. Switched approaches in ab367236a01: the prior pass unconditionally replaced every backslash-n in every context string, which would have mangled regex patterns and any other literal-backslash-n payload. The fix now lives in PrintableDict.__str__/__repr__ — when a string contains an actual newline (or CR/tab) it is emitted as a YAML double-quoted scalar via yaml.safe_dump, otherwise repr() is used unchanged. A literal backslash-n in source data has no real newline, so it goes through repr() and remains 'foo\\nbar' (Python single-quoted), which YAML loads back as 'foo\\nbar'.

dwoz replied:
Switched approaches — the new fix in ab367236a01 triggers on any of '\\n', '\\r', or '\\t' in the source string and rewrites it as a YAML double-quoted scalar, so CRLF in import_yaml data round-trips correctly on Windows too. os.linesep isn't quite the right axis here — the corruption was driven by what Python repr() escapes (which is platform-independent), not by the host's line separator.

dwoz committed:
Rework #30690 fix: emit YAML-safe scalars in PrintableDict

Drop the indiscriminate _unescape_newlines() pass in render_jinja_tmpl
that silently replaced every '\n' (backslash-n) with a real newline
across the entire context. That broke any string legitimately
containing a literal backslash-n (regex patterns, escape sequences,
etc.) and missed CRLF / tab cases on Windows.

Instead, fix the actual source of the corruption: PrintableDict.__str__
and __repr__ emitted string values via Python repr(), which turns a
real newline in the source data into the two-character sequence \n.
When such output is interpolated into a YAML state file via
{{ some_dict }}, the downstream YAML parser sees a literal backslash-n
rather than a newline.

PrintableDict now emits string values containing newline / CR / tab as
YAML double-quoted scalars (via yaml.safe_dump) which the YAML parser
decodes back to the original characters. Strings without those
characters continue to use repr() for backward compatibility, so
literal backslash-n in source data is preserved unchanged.

Also revert the unrelated OrderedDict import switch to salt.utils.odict
(which is deprecated) and re-import from collections, drop the unused
salt.utils.files import in the test, and switch the integration test
to tmp_path so it runs on Windows.

Refs: review comments from @bdrx312 and @twangboy on PR #68535

dwoz commented on salt/utils/templates.py:
```diff
@@ -33,6 +32,7 @@
 from salt.loader.context import NamedLoaderContext
 from salt.loader.dunder import __file_client__
 from salt.utils.decorators.jinja import JinjaFilter, JinjaGlobal, JinjaTest
+from salt.utils.odict import OrderedDict
```
Reverted in ab367236a01, back to 'from collections import OrderedDict'. The salt.utils.odict switch was an unrelated accident.

dwoz committed:
Drop 'mode' from test_issue_30690 SLS for Windows

file.managed rejects '- mode: 644' on Windows with 'The mode option
is not supported on Windows', failing the state run before any file
gets written. The test only validates rendered content, so mode is
not required.

Commit:
Normalize CRLF in test_issue_30690 content comparison

rstrip('\n') leaves stray '\r' characters in the trailing whitespace
on Windows, where file.managed writes CRLF. Normalize CRLF to LF then
rstrip() so the import_yaml vs direct-context content equality check
is platform-independent. The 'no literal backslash-n' assertion above
remains the primary check for the bug under test.
